### ISSUE TITLE
Receiving 'body used already for: https://app.useanvil.com/graphql' exception when calling requestGraphQL

### ISSUE THREAD

**Issue by ctmckenna** [anvilco/node-anvil#376]
## Expected Behavior
If requestGraphQL throws an exception it should be done intentionally rather than from calling json() twice

## Actual Behavior
requestGraphQL first calls json() [here](https://github.com/anvilco/node-anvil/blob/main/src/index.js#L673).  I believe this call to json sometimes throws an exception which is caught but since no state is changed and function continues with an undefined json variable, json() is called again [here](https://github.com/anvilco/node-anvil/blob/main/src/index.js#L700).  For the second json to be called, the statusCode would have to be < 300.

## Steps to Reproduce the Problem

I don't know the circumstances that leads the first json call to throw an exception since we haven't had the debug variable set, but that's the only code path i can see where the 'body used already' error may arise.  I've just enabled the debug variable so will hopefully get more context soon.  But based on the code it seems we would have to receive a 200 response with an application/json content type and invalid json

**Comment by newhouse** [comment]
Hello @ctmckenna and thanks for using Anvil's Node client.

Could you provide the details of the error output? It would be nice to also see what is revealed with the `debug` option enabled on the client.

What is the call that you're making? I will try to replicate this

**Comment by ctmckenna** [comment]
@newhouse, I just enabled the debug option.  I'll follow up with more details next time it happens with what gets logged there.

In the meantime, it happened on September 19th at 12:55:53 PM PDT when calling the weld query endpoint:
![Screenshot 2023-09-26 at 12 22 52 PM](https://github.com/anvilco/node-anvil/assets/1254699/7a85a5c5-a251-4688-af4b-57ff54c39eda)

September 21st at 9:50:16am PDT when calling the weldData query endpoint
And latest was September 24th at 10:24:11am PDT when calling the weldData query endpoint

The parameters of those queries aren't quickly accessible but i could probably find them if needed

**Comment by newhouse** [comment]
@ctmckenna 
Can you paste in the raw stack trace, please? There are 4 more frames hidden in your Sentry snapshot?  I'm just trying to see the full stack trace and info.

Here's an example of how to get at it:
![image](https://github.com/anvilco/node-anvil/assets/2524009/2da46b90-3466-419f-bc66-4f58728d3f7a)

**Comment by ctmckenna** [comment]
<img width="919" alt="Screenshot 2023-09-26 at 7 50 08 PM" src="https://github.com/anvilco/node-anvil/assets/1254699/159b0730-84ff-4bed-9778-6fa05623a2c9">

**Comment by clickclickonsal** [comment]
@newhouse I'm on @ctmckenna's team. I'm seeing a premature close error right before the "body already used for" error

See stacktrace

```

FetchError: Invalid response body while trying to fetch https://app.useanvil.com/graphql: Premature close
    at consumeBody (file:///app/node_modules/@anvilco/anvil/node_modules/node-fetch/src/body.js:210:60)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Response.json (file:///app/node_modules/@anvilco/anvil/node_modules/node-fetch/src/body.js:119:18)
    at async /app/node_modules/@anvilco/anvil/dist/index.js:608:18
    at async Anvil.requestGraphQL (/app/node_modules/@anvilco/anvil/dist/index.js:494:9)
    at async requestGraphQL (/app/dist/lib/server/anvilClient/requestGraphQL.js:33:42)
    at async AnvilClient.fetchWeld (/app/dist/lib/server/anvilClient/anvilClient.js:51:26)
    at async Object.service [as originalProvider] (/app/dist/ddd/WorkforceManagement/Anvil/application/service/generateAnvilLink.js:66:22)
  type: 'system',
  errno: 'ERR_STREAM_PREMATURE_CLOSE',
  code: 'ERR_STREAM_PREMATURE_CLOSE',
  erroredSysCall: undefined
}
Using statusText instead
{"level":"log","message":"API Request Failed:","dd":{"trace_id":"9069754729766566701","span_id":"1389218408148981733","service":"seso-app","version":"v8885","env":"prod"}} {
  error: {
    message: 'body used already for: https://app.useanvil.com/graphql',
    name: 'TypeError',
    cause: undefined,
    stackTrace: 'TypeError: body used already for: https://app.useanvil.com/graphql\n' +
      '    at consumeBody (file:///app/node_modules/@anvilco/anvil/node_modules/node-fetch/src/body.js:162:9)\n' +
      '    at Response.json (file:///app/node_modules/@anvilco/anvil/node_modules/node-fetch/src/body.js:119:24)\n' +
      '    at /app/node_modules/@anvilco/anvil/dist/index.js:635:42\n' +
      '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
      '    at async Anvil.requestGraphQL (/app/node_modules/@anvilco/anvil/dist/index.js:494:9)\n' +
      '    at async requestGraphQL (/app/dist/lib/server/anvilClient/requestGraphQL.js:33:42)\n' +
      '    at async AnvilClient.fetchWeld (/app/dist/lib/server/anvilClient/anvilClient.js:51:26)\n' +
      '    at async Object.service [as originalProvider] (/app/dist/ddd/WorkforceManagement/Anvil/application/service/generateAnvilLink.js:66:22)\n' +
  },
  request: {
    method: 'GET',
    originalUrl: '/api/onboarding-endpoint/REDACTED_UUID?userUuid=ANOTHER_REDACTED_UUID',
    url: '/api/onboarding-endpoint/REDACTED_UUID?userUuid=ANOTHER_REDACTED_UUID',
     body: {}
  },
  response: { statusCode: 500 }
```

**Comment by newhouse** [comment]
@clickclickonsal @ctmckenna 

What it seems like is that whatever call you're making is triggering a bad response from our API (which is probably the `Premature close` error), and then our Node client is also making the mistake of trying to read the response body again (which is the subsequent `body used already for` error).

Seems like you guys are able to reproduce/trigger this error somewhat easily/regularly.

Can you give me the information I need to replicate this? You can leave out your API key (of course) but if I could replicate it then I'm sure I can figure out what's going on and how to fix it. What calls are you making to our Node client in your `fetchWeld` method, and to our `requestGraphQL` method?

Does it error consistently with the same options provided, or is it intermittent?

Also, you can send an e-mail to `support@useeanvil.com` if you want to continue this discussion via support-email.

**Comment by ctmckenna** [comment]
I just sent an email to support@useanvil.com with more details.  I can follow up with some more details shortly as well

**Comment by wi-ski** [comment]
Loom because everyone loves a loom: https://www.loom.com/share/f2f0bd7734e74ed3a7fa6e8f9a37e990?sid=971442dd-615b-444b-8e9b-b5d3c26c11ee

**Comment by wi-ski** [comment]
@newhouse Please see the PR, it should at least convey the idea.

**Comment by newhouse** [comment]
@wi-ski @clickclickonsal @ctmckenna 

I'm doing to dive into this in the next hour or so and will give you updates and/or get it fixed! Thanks for all the details, help is on the way 🪂

**Comment by newhouse** [comment]
OK, so there are sorta 2 things to work on here:

1. What is with the "premature close" thing? Right now, I don't know yet but that's where I'm going to focus next.
2. The handling of errors in the spot in question: handle it and return the original Error. For this I have a fix candidate published now as [`3.2.0-beta.0`](https://www.npmjs.com/package/@anvilco/anvil/v/3.2.0-beta.0)

Can someone install and try [`3.2.0-beta.0`](https://www.npmjs.com/package/@anvilco/anvil/v/3.2.0-beta.0) and let me know if that at least fixes the error handling behavior?

@wi-ski @ctmckenna @clickclickonsal

**Comment by newhouse** [comment]
I have also figured out, I believe, the underlying bug for the `Premature close`...will get a fix out for both of those today.

**Comment by newhouse** [comment]
OK @ctmckenna @clickclickonsal @wi-ski 

This should hopefully all be fixed in [`3.2.0-beta.1`](https://www.npmjs.com/package/@anvilco/anvil/v/3.2.0-beta.1)

Please try it out and let me know...if I don't hear from you I'm going to assume we are all good and close this soon.

**Comment by newhouse** [comment]
Published officiall as `3.2.0`.

Closing this issue for now
